### PR TITLE
fix: handles when main file is inside a package

### PR DESCRIPTION
**Pull request by tivrfoa**
PR https://github.com/jbangdev/jbang/pull/469 introduced ability to search for `//SOURCES` and `//DEPS` in multiple files.
But there is one case (at least) where it doesn't work:
  1. main file is a local file that has `package` declared;
  2. It references other file(s) in `//SOURCES`  that is not in the same or child package.

Then it is not able to find the classes in the other packages because it tries to resolve their paths using the base path of the main file.

This PR does two things:
  1. it takes into account the package name and sets --source-path accordingly. eg:
      jbang /home/user/scripts/com/company/Main.java
      then --source-path will be /home/user/scripts
    This change should help cases where `//SOURCES` is not used, so javac will be able to resolve the paths. *ps: by not using `//SOURCES` you would not be able to use some of the JBang features.*
  2. it correctly resolves base path for `//SOURCES` when main file is in a package.

~~This PR does not handle `//SOURCES` with absolute path to local files. And relative paths only work if it's the same level or a sublevel of the main file.~~
The code below already works in JBang!! That is really cool!
```java
//SOURCES ../../../b.java
//SOURCES /home/leandro/dev/java/jbang/c.java

public class oi {
    
    public static void main(String args[]) {
        System.out.println("oi");
		b.say();
		c.say();
    }
}
```
```shell
leandro:distributions$ pwd
/home/leandro/dev/githubprojects/tmp-jbang/jbang/build/distributions
leandro:distributions$ rm -fR ~/.jbang/cache/jars/oi.java.*
leandro:distributions$ ./jbang/bin/jbang oi.java 
[jbang] Building jar...
oi
I am b
I am c
```

### CODE REVIEW

**Comment by maxandersen** [comment]
Why do you think this fixes 464 with pattern matching ? The //sources is needed by edit and of classes not reachable by main class.

**Comment by tivrfoa** [comment]
> Why do you think this fixes 464 with pattern matching ? The //sources is needed by edit and of classes not reachable by main class.

oh. I thought he just wanted to run. I updated to should *help* =)

So, on my machine (Ubuntu) all tests passed, but failed CI Ubuntu =(

**Comment by maxandersen** [comment]
failure looks like github action infrastructure fail. restarted the jobs.

**Comment by tivrfoa** [comment]
> failure looks like github action infrastructure fail. restarted the jobs.

This PR is a little big because I added the test that you didn't merge from PR #469 
I still don't know why CI is failing here, but works on my machine (Ubuntu)

**Comment by tivrfoa** [comment]
> I still don't know why CI is failing here, but works on my machine (Ubuntu)

It didn't work because `--source-path` was introduced in Java 9.
So now I check the Java version, and if <= 8, then `-sourcepath` is used.

**Comment by quintesse** [comment]
> So now I check the Java version, and if <= 8, then -sourcepath is used.

Why check? Why not always use `-sourcepath`?

**Comment by tivrfoa** [comment]
> Why check? Why not always use `-sourcepath`?

Thank you Tako!
I thought the old way was removed, but from Java 9 both -sourcepath and --source-path are supported.

File: src/main/java/dev/jbang/Script.java [hunk 1]
```diff
@@ -74,6 +74,7 @@
 	private List<KeyValue> agentOptions;
 	private String preMainClass;
 	private String agentMainClass;
+
 	/**
```
Comment (maxandersen) [inline src/main/java/dev/jbang/Script.java:78]: this looks like an unnecessary change?

File: src/main/java/dev/jbang/ScriptResource.java [hunk 2]
```diff
@@ -13,11 +15,19 @@
...
+	// eg: /home/user/scripts/org/acme/Main.java, then
+	// originalSourceFileBaseDir = /home/user/scripts
+	private final File originalSourceFileBaseDir;
 
 	public ScriptResource(String scriptURL, File urlCache, File file) {
 		this.originalResource = scriptURL;
 		this.sourceCacheDir = urlCache;
 		this.file = file;
+		if (Util.isJavaFile(scriptURL))
+			this.originalSourceFileBaseDir = setOriginalSourceFileBaseDir(scriptURL);
```
Comment (maxandersen) [inline src/main/java/dev/jbang/ScriptResource.java:28]: this reads wrong - why call setOriginalSourceFileBaseDir(scriptUrl) to get the return value ...why is this not just calculated (and cached) in `getOriginalSourceFileBaseDir()` ?

File: src/main/java/dev/jbang/cli/BaseScriptCommand.java [hunk 3]
```diff
@@ -155,49 +155,50 @@ public static Script prepareScript(String scriptResource, List<String> arguments
...
-
-			s.setResolvedSources(resolvedSourcePaths);
+			s.setResolvedSources(resolveSOURCESRecursively(s));
 		} catch (FileNotFoundException e) {
 			throw new ExitException(1, e);
 		}
 		return s;
 	}
 
+	public static List<Path> resolveSOURCESRecursively(Script script) throws IOException {
```
Comment (maxandersen) [inline src/main/java/dev/jbang/cli/BaseScriptCommand.java:165]: this does not seem related to this fix ?

File: src/main/java/dev/jbang/ScriptResource.java [hunk 4]
```diff
@@ -43,19 +53,49 @@ public Path fetchIfNeeded(String resource) {
...
+				if (pkgLine.length == 1)
+					continue;
+				String packageName = pkgLine[1];
+				packageName = packageName.split(";")[0].trim(); // remove ';'
+				String pkg = packageName.replace(".", File.separator); // a.b.c -> linux a/b/c, windows a\b\c
+				String pkgFile = pkg + File.separator + Util.getFileName(scriptURL);
+				return new File(scriptURL.substring(0, scriptURL.lastIndexOf(pkgFile)));
+			}
+		} catch (FileNotFoundException e) {
+			throw new RuntimeException(e);
```
Comment (maxandersen) [inline src/main/java/dev/jbang/ScriptResource.java:92]: I would argue this methods must never fail. It can warn about some mismatched dir patterns but it should not fail hard.

File: src/main/java/dev/jbang/cli/BaseScriptCommand.java (hunk 3 above)
Comment (tivrfoa) [reply src/main/java/dev/jbang/cli/BaseScriptCommand.java:165]: This one you didn't merge from previous PR. You also didn't merge the test.
I extracted to a function to make it easier to test. But it end up making the code more readable in my opinion.

**Comment by maxandersen** [comment]
> This PR is a little big because I added the test that you didn't merge from PR #469

if I missed sometihng there lets put that in separate PR if possible. lets not make a habit of mixing two concerns in one PR :)

**Comment by tivrfoa** [comment]
I'm closing this PR, as it seems everything just works if JBang is executed from the correct folder.
